File: returned/python/today6/largest_rectangle_in_histogram.py
# https://leetcode.com/problems/largest-rectangle-in-histogram/description/
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        largest = float('-inf')
        stack = []

        for i, h in enumerate(heights):
            if not stack or h >= stack[-1][1]:
                stack.append((i, h))
            else:
                while stack and stack[-1][1] > h:
                    j, height = stack.pop()
                    area = (i - j) * height
                    if area > largest:
                        largest = area
                stack.append((j, h))
 
        while stack:
            j, height = stack.pop()
            area = (len(heights) - j) * height
            if area > largest:
                largest = area

        return largest

File: csharp/Leet/Today2/LargestRectangleInHistogram.cs
// https://leetcode.com/problems/largest-rectangle-in-histogram/description/
namespace Today2.LargestRectangleInHistogram;

public class Solution
{
    public int LargestRectangleArea(int[] heights)
    {
        var result = 0;
        var stack = new Stack<(int Index, int Height)>();
        stack.Push((-1, 0));
        for (int i = 0; i < heights.Length; i++)
        {
            var lastIndex = i;
            while (heights[i] < stack.Peek().Height)
            {
                var (index, height) = stack.Pop();
                var area = (i - index) * height;
                result = Math.Max(result, area);
                lastIndex = index;
            }
            stack.Push((lastIndex, heights[i]));
        }

        while (stack.Count > 0)
        {
            var (index, height) = stack.Pop();
            var area = (heights.Length - index) * height;
            result = Math.Max(result, area);
        }

        return result;
    }
}